config/desc/imsim/readout.py

```diff
@@ -26,64 +223,61 @@ def finalize(self, config, base, main_data, logger):
+        wcs = main_data[0].wcs
+        crpix1, crpix2 = wcs.crpix
+        hdus = fits.HDUList(get_primary_hdu(config, base, main_data))
         for amp_num, amp in enumerate(amps):
-            amp_name = det_name + "-C%02d"%amp_num  # XXX: Pretty sure this isn't right.
-            logger.warning("Amp %s has bounds %s.",amp_name,amp.bounds)
-
-            amp.header = galsim.FitsHeader()
-            amp.header['AMP_NAME'] = amp_name
-            # TODO: Add other header information.
-
-            # XXX: I think these probably all should have their origin set to 1,1?
-            #      At this point, the amp images still ahve the bounds from the original
-            #      eimage, so we would need this next line to get back to 1,1 as the origin.
-            amp.setOrigin(1,1)
-
-            # TODO: I think probably also the WCS might need to be changed?  Not sure.
-            #       At this point, the WCS should be accurate for the re-origined image.
-            #       But if we do something more complicated like mirror image the array or
-            #       something like that, the WCS would probably need to change.
-
-        return amps
+            channel = 'C' + channels[amp_num]
+            amp_info = ccd_readout.ccd[channel]
+            raw_data_bounds = amp_info.raw_data_bounds
+            hdu = fits.CompImageHDU(np.array(amp.array, dtype=np.int32),
+                                    compression_type='RICE_1')
+            wcs.writeToFitsHeader(hdu.header, main_data[0].bounds)
+            hdu.header['EXTNAME'] = 'Segment' + channels[amp_num]
+            xsign = -1 if amp_info.raw_flip_x else 1
+            ysign = -1 if amp_info.raw_flip_y else 1
+            height, width = raw_data_bounds.numpyShape()
+            hdu.header['CRPIX1'] = xsign*crpix1 + x_seg_offset[amp_num]*width
+            hdu.header['CRPIX2'] = ysign*crpix2 + y_seg_offset[amp_num]*height
+            hdu.header['CD1_2'] *= -xsign
+            hdu.header['CD2_2'] *= -xsign
+            hdu.header['CD1_1'] *= -ysign
+            hdu.header['CD2_1'] *= -ysign
+            hdu.header['DATASEC'] = section_keyword(raw_data_bounds)
+            hdu.header['DETSEC'] = section_keyword(amp_info.bounds,
+                                                   flipx=amp_info.raw_flip_x,
+                                                   flipy=amp_info.raw_flip_y)
+            hdus.append(hdu)
+            amp_name = '_'.join((det_name, channel))
+            logger.warning("Amp %s has bounds %s.", amp_name,
+                           hdu.header['DETSEC'])
+        return hdus
 
     def writeFile(self, file_name, config, base, logger):
         """Write this output object to a file.
 
-        The base class implementation is appropriate for the cas that the result of finalize
-        is a list of images to be written to a FITS file.
+        The base class implementation is appropriate for the cas that
```

Comment:
Typo (mine I suspect) cas -> case.
But in fact, this comment is not really relevant anymore, since the output from your `finalize` is now an `HDUList`.  So probably best to just remove this paragraph.